source/src/BNFC/Backend/Haskell/CFtoHappy.hs

```diff
@@ -231,36 +257,65 @@ footer = unlines $
-  "Char"    -> "L_charac { PT _ (TC $$) }"
-  own       -> "L_" ++ own ++ " { PT _ (T_" ++ own ++ " " ++ posn ++ ") }"
-    where posn = if isPositionCat cf own then "_" else "$$"
+specialToks :: CF -> Bool -> [String]
+specialToks cf functor = (`map` literals cf) $ \t -> case t of
+  "Ident"   -> "L_Ident  { PT _ (TV " ++ posn t ++ ") }"
+  "String"  -> "L_quoted { PT _ (TL " ++ posn t ++ ") }"
+  "Integer" -> "L_integ  { PT _ (TI " ++ posn t ++ ") }"
+  "Double"  -> "L_doubl  { PT _ (TD " ++ posn t ++ ") }"
+  "Char"    -> "L_charac { PT _ (TC " ++ posn t ++ ") }"
+  own       -> "L_" ++ own ++ " { PT _ (T_" ++ own ++ " " ++ posn own ++ ") }"
+  where
+    posn tokenCat = if isPositionCat cf tokenCat || functor then "_" else "$$"
 
-specialRules :: ModuleName -> TokenText -> CF -> String
-specialRules absName tokenText cf = unlines . intersperse "" . (`map` literals cf) $ \case
+specialRules :: ModuleName -> Bool -> TokenText -> CF -> String
+specialRules absName functor tokenText cf = unlines . intersperse "" . (`map` literals cf) $ \t -> case t of
     -- "Ident"   -> "Ident   :: { Ident }"
     --         ++++ "Ident    : L_ident  { Ident $1 }"
-    "String"  -> "String  :: { String }"
-            ++++ "String   : L_quoted { " ++ stringUnpack "$1" ++ " }"
-    "Integer" -> "Integer :: { Integer }"
-            ++++ "Integer  : L_integ  { (read (" ++ stringUnpack "$1" ++ ")) :: Integer }"
-    "Double"  -> "Double  :: { Double }"
-            ++++ "Double   : L_doubl  { (read (" ++ stringUnpack "$1" ++ ")) :: Double }"
-    "Char"    -> "Char    :: { Char }"
-            ++++ "Char     : L_charac { (read (" ++ stringUnpack "$1" ++ ")) :: Char }"
-    own       -> own ++ " :: { " ++ qualify own ++ "}"
-            ++++ own ++ "  : L_" ++ own ++ " { " ++ qualify own ++ posn ++ " }"
-      where posn = if isPositionCat cf own then " (mkPosToken $1)" else " $1"
+    "String"  -> "String  :: { " ++ mkTypePart t ++ " }"
+            ++++ "String   : L_quoted { " ++ mkBodyPart t ++ " }"
+    "Integer" -> "Integer :: { " ++ mkTypePart t ++ " }"
+            ++++ "Integer  : L_integ  { " ++ mkBodyPart t ++ " }"
+    "Double"  -> "Double  :: { " ++ mkTypePart t ++ " }"
+            ++++ "Double   : L_doubl  { " ++ mkBodyPart t ++ " }"
+    "Char"    -> "Char    :: { " ++ mkTypePart t ++ " }"
+            ++++ "Char     : L_charac { " ++ mkBodyPart t ++ " }"
+    own       -> own ++ " :: { " ++ mkTypePart (qualify own) ++ " }"
+            ++++ own ++ "  : L_" ++ own ++ " { " ++ mkBodyPart t ++ " }"
   where
+    mkTypePart tokenCat = if functor then "(Maybe (Int, Int), " ++ tokenCat ++ ")" else tokenCat
+    mkBodyPart tokenCat
+      | null mkPosPart = mkValPart tokenCat
+      | otherwise      = "(" ++ mkPosPart ++ ", " ++ mkValPart tokenCat ++ ")"
+    mkPosPart = if functor then "Just (tokenLineCol $1)" else ""
+    mkValPart tokenCat =
+      case tokenCat of
+        "String"  -> if functor then stringUnpack "(tokenText $1)"
```

Comment:
Testcase `235_SymbolsOverlapTokens` is broken: `String` literals now arrive quoted in the abstract syntax.  This is due to the use of `tokenText` here, I suppose.  Note that  the generated `tokenText` for `String` tokens uses the `show` function, which adds a level of quotation.

Reply:
That is my mistake. Indeed, the `show` function in `tokenTest` adds an extra level of quotation to `String` literals. 

The simplest and clearest way I can find out is to replace the `tokenText` function with a partial one `(\(PT _ (TL s)) -> s)`:

```haskell
"String"  -> if functor then stringUnpack "(\\(PT _ (TL s)) -> s) $1" 
                        else stringUnpack "$1"
```

Although the function is partial, there will never be a `Tok` other than `TL s` when `"String"` is matched. So it is safe, I think. 

(And I finally find out why my testing gave me unexpected results because I forgot to replace `bnfc` with the latest build :( )